### PR TITLE
doc: minimal documentation of supported platforms

### PR DESCRIPTION
**Pull request by asymmetric**
This commit adds minimal documentation of the supported platforms.

More exhaustive documentation would require producing a list of platforms for each of the 7 tiers. This was attempted in #245368, but it quickly became clear that that would be a long-term effort.
    
In the meantime, this commit adds the most important information to the manual.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by henrik-ch — APPROVED** [review]
lgtm

**Review by 06kellyjac — APPROVED** [review]
LGTM

Might be worth getting an updated platform list at some point

File: doc/platform-support.appendix.md
Comment (infinisil) [inline doc/platform-support.appendix.md:6]:
```suggestion
- `x86_64-linux`: Highest level of support, both in terms of CI and of developer attention.
- `aarch64-linux`: Well supported, with most packages building successfully in CI.
- `aarch64-darwin`: Receives better support over `x86_64-darwin`
- `x86_64-darwin`: Receives some support
```

File: doc/platform-support.appendix.md
Comment (infinisil) [inline doc/platform-support.appendix.md:8]:
```suggestion
A more formal definition of different platform support tiers was accepted with [RFC046], though it has not been fully implemented so far.
```

File: doc/manual.md.in
Comment (infinisil) [inline doc/manual.md.in:19]: I feel like this really shouldn't need to be in an appendix. This is fairly important information and should probably be somewhere in the beginning of the manual. Could you look around to see if it would fit somewhere else?

**Review by infinisil — DISMISSED** [review]
I feel like this is missing a link to Hydra, but it's also really difficult to establish a good such link, because the [nixpkgs Hydra project](https://hydra.nixos.org/project/nixpkgs) is not great to look at. Ideally we'd be able to express platform support as a percentage of packages that build successfully. But that's for the future.

File: doc/manual.md.in
Comment (asymmetric) [reply doc/manual.md.in:19]: Right, although it's RFC046 that stated this should be an appendix:

> The definition of support tiers and a list of platforms with the corresponding support tiers is added as an appendix to the Nixpkgs manual. 

Not really sure how binding that is. And yes, we're not exactly implementing that RFC here, but assuming one day it will be implemented (unlikely IMO), and assuming it's binding, then we would have to relocate the section to an appendix, which would be bad.

This also makes me wonder: if we expect the content to move, then the anchor should not contain the preifx `appendix`.

File: doc/platform-support.appendix.md
Comment (asymmetric) [reply doc/platform-support.appendix.md:8]: it's not clear what hasn't been implemented. I don't want readers to have to figure that out by reading the whole RFC. That's why I mentioned specifically the platform list.

File: doc/platform-support.appendix.md
Comment (asymmetric) [reply doc/platform-support.appendix.md:6]: Do you find this better? I personally prefer the more discursive style, in this case.

**Comment by asymmetric** [comment]
@infinisil How strongly do you feel about your requests for changes? As it is, they are all part of a blocking review. For stylistic issues, it would be great if we could either:

- Leave it up to the author
- Justify with some consensus (ideally written down)
- Make it clear that they are "nits"/personal opinion

File: doc/platform-support.appendix.md
Comment (06kellyjac) [reply doc/platform-support.appendix.md:6]: Sometimes you do just want to scan through.

Could have the more verbose description and a more concise table showing support?

Comment (7c6f434c) [reply doc/platform-support.appendix.md:6]: Also the end state is probably a full table (eventually), so we can start in this direction anyway?

File: doc/manual.md.in
Comment (7c6f434c) [reply doc/manual.md.in:19]: Maybe have a short description in the main body, and copy the tier definitions into appendix?

**Comment by 7c6f434c** [comment]
> Ideally we'd be able to express platform support as a percentage of packages that build successfully.

More like percentages of some well-defined core package sets, so that technical trade-offs about building `python3*Packages` don't change the stats when not warranted.

File: doc/platform-support.appendix.md
Comment (7c6f434c) [inline doc/platform-support.appendix.md:6]: Isn't `x86_64-darwin` and `aarch64-darwin` situation more complicated than really comparable levels because from time to time CI before merging actually works better for the former?

Comment (asymmetric) [reply doc/platform-support.appendix.md:6]: Sounds good, I'll switch to a table.

File: doc/platform-support.appendix.md
Comment (fricklerhandwerk) [reply doc/platform-support.appendix.md:8]:
```suggestion
A more formal definition of different platform support tiers was accepted with [RFC046], but is not fully implemented so far.
The preliminary platform list in the RFC's [Appendix A] can be used as guidance, but is not up to date.
```

File: doc/manual.md.in
Comment (fricklerhandwerk) [reply doc/manual.md.in:19]: Yes, support information should be in one of the first sections in the manual.

File: doc/platform-support.appendix.md
Comment (fricklerhandwerk) [reply doc/platform-support.appendix.md]: @toonn may have a thought on this.

Comment (toonn) [reply doc/platform-support.appendix.md]: I can't really disagree with the statement. Most of NixOS/darwin-maintainers seem to be aarch64-darwin users. The Darwin builders for Hydra are aarch64-darwin for the most part, emulating x86_64-darwin when necessary. Maybe a qualifier like "slightly" is appropriate. The goal is for them to be equal for the foreseeable future.

File: doc/platform-support.appendix.md
Comment (infinisil) [reply doc/platform-support.appendix.md:8]: That sounds good to me

File: doc/manual.md.in
Comment (infinisil) [reply doc/manual.md.in:19]: Yeah, and for now we don't need any lengthy details. So I think this entire section should just be somewhere in the beginning of the manual instead.

**Comment by infinisil** [comment]
> @infinisil How strongly do you feel about your requests for changes? As it is, they are all part of a blocking review.

The stylistic changes I don't care much about, but https://github.com/NixOS/nixpkgs/pull/254741#discussion_r1323679237 is more important.

File: doc/manual.md.in
Comment (asymmetric) [reply doc/manual.md.in:19]: Sounds good, I'll look for a place. I do think we need to make this section a little more discursive if it's not tucked in in an appendix. Don't worry, I'll leave the table in ;)

**Comment by asymmetric** [comment]
Please review again. I've moved the chapter way up in the manual (in the position we discussed in the meeting).

I've decided not to add an appendix with an outdated platform list -- this will have to be done in a follow up PR, if someone has the resources to produce an up to date one.

**Comment by asymmetric** [comment]
@pennae As someone who worked with the manual a lot: does it make sense that I introduced a chapter with no sections? Or should it be a section instead?

**Comment by asymmetric** [comment]
@pennae the consensus is that it should be prominent, so not an appendix. If I made it an inline content of [Using Nixpkgs](https://nixos.org/manual/nixpkgs/stable/#part-using), could I give it an anchor?

**Comment by 7c6f434c** [comment]
> I've decided not to add an appendix with an outdated platform list -- this will have to be done in a follow up PR, if someone has the resources to produce an up to date one.

But at least the definitions could be copied to an appendix?

**Comment by asymmetric** [comment]
> But at least the definitions of the tiers could be copied to an appendix?

@7c6f434c my concern with doing that is: I think that unless the tiers are actually implemented[^1] in practice, there is little point in adding them to the manual, and could even be misleading.

It seems to me that there's at least some (and maybe a lot of) work that would need to be done "behind the scenes" before the tiers can be said to represent something actually existing. At that point the definitions should definitely be added to the manual.

But that's work I realized I wasn't able to do (see https://github.com/NixOS/nixpkgs/pull/245368), so I'd prefer to keep these two aspects separated.

[^1]: By implemented I mean: followed by platform maintainers, reflected in organizational structure, CI
[^2]: It could also be that I'm wrong, and that the tiers are mostly implemented, which goes to show I'm not the right person to make the call of whether this should be included.

**Comment by asymmetric** [comment]
> I feel like this is missing a link to Hydra, but it's also really difficult to establish a good such link, because the [nixpkgs Hydra project](https://hydra.nixos.org/project/nixpkgs) is not great to look at.

@infinisil there's a link to Hydra in the [previous section](https://nixos.org/manual/nixpkgs/stable/#overview-of-nixpkgs), and tellingly, it points to the GitHub repo.

**Comment by 7c6f434c** [comment]
> By implemented I mean: followed by platform maintainers, reflected in organizational structure, CI.

It kind of has nothing to reflect in CI, it works the other way round, no? At the end of the day, tiers are rules of arbitration of tradeoffs when platform maintainers and package maintainers have different interests. The rules take into account the CI situation.

**Review by fricklerhandwerk — APPROVED** [review]
@henrik-ch and me checked what it looks like, and we think it's fine having it as a top-level chapter. Some small wording suggestions, otherwise good to merge.

File: doc/using/platform-support.chapter.md
Comment (asymmetric) [reply doc/using/platform-support.chapter.md:11]: Thanks, I was looking for a way out of that one :D

File: doc/using/platform-support.chapter.md
Comment (infinisil) [inline doc/using/platform-support.chapter.md:4]:
```suggestion
Here is the list of platforms with some support:
```

**Review by infinisil — APPROVED** [review]
Otherwise this looks good!

**Review by infinisil — APPROVED** [review]

**Review by infinisil — APPROVED** [review]

File: doc/using/platform-support.chapter.md
Comment (infinisil) [reply doc/using/platform-support.chapter.md]: Either sounds good!

**Comment by infinisil** [comment]
Did you want to squash the commits yourself? Otherwise I'll just squash them when merging

**Comment by asymmetric** [comment]
Yes, it was my intention to fixup before merging myself :)

**Comment by asymmetric** [comment]
Can't be backported because 23.05 still uses the XML setup for the manuals.

**Comment by infinisil** [comment]
As also mentioned [on Matrix](https://matrix.to/#/!avYyleMexqjFHoqrME:nixos.org/$SVifc-el8UNBMapvF-oaC6tYRdvtNQ6FwkHAJq1fJms?via=nixos.org&via=matrix.org&via=nixos.dev), a manual backport would still be doable, however it's not very easy and this doesn't seem worth it, so not crucial.

**Pull request merged**
